reactor-netty-http/src/main/java/reactor/netty/http/client/SpnegoAuthProvider.java

```diff
@@ -0,0 +1,315 @@
+	private final AtomicInteger retryCount = new AtomicInteger(0);
+	private static final int MAX_RETRY_COUNT = 1;
+
+	/**
+	 * Constructs a new SpnegoAuthProvider with the specified configuration.
+	 * <p>
+	 * This constructor is private and should only be used by the {@link Builder}.
+	 * Use {@link #builder(SpnegoAuthenticator)} to create instances.
+	 * </p>
+	 *
+	 * @param authenticator the authenticator to use for SPNEGO authentication (must not be null)
+	 * @param unauthorizedStatusCode the HTTP status code that indicates authentication failure
+	 * @param serviceName the service name for constructing service principal names
+	 * @param resolveCanonicalHostname whether to resolve canonical hostnames for service principals
+	 */
+	private SpnegoAuthProvider(SpnegoAuthenticator authenticator, int unauthorizedStatusCode, String serviceName, boolean resolveCanonicalHostname) {
+		this.authenticator = authenticator;
+		this.unauthorizedStatusCode = unauthorizedStatusCode;
+		this.serviceName = serviceName;
+		this.resolveCanonicalHostname = resolveCanonicalHostname;
+	}
+
+	/**
+	 * Creates a new builder for configuring SPNEGO authentication provider.
+	 *
+	 * @param authenticator the authenticator to use for SPNEGO authentication
+	 * @return a new builder instance
+	 */
+	public static Builder builder(SpnegoAuthenticator authenticator) {
+		return new Builder(authenticator);
+	}
+
+	/**
+	 * Builder for creating SpnegoAuthProvider instances.
+	 */
+	public static final class Builder {
+		private final SpnegoAuthenticator authenticator;
+		private int unauthorizedStatusCode = 401;
+		private String serviceName = "HTTP";
+		private boolean resolveCanonicalHostname;
+
+		private Builder(SpnegoAuthenticator authenticator) {
+			this.authenticator = authenticator;
+		}
+
+		/**
+		 * Sets the HTTP status code that indicates authentication failure.
+		 *
+		 * @param statusCode the status code (default: 401)
+		 * @return this builder
+		 */
+		public Builder unauthorizedStatusCode(int statusCode) {
+			this.unauthorizedStatusCode = statusCode;
+			return this;
+		}
+
+		/**
+		 * Sets the service name for the service principal.
+		 *
+		 * @param serviceName the service name (default: "HTTP")
+		 * @return this builder
+		 */
+		public Builder serviceName(String serviceName) {
+			this.serviceName = serviceName;
+			return this;
+		}
+
+		/**
+		 * Sets whether to resolve canonical hostname.
+		 *
+		 * @param resolveCanonicalHostname true to resolve canonical hostname (default: false)
+		 * @return this builder
+		 */
+		public Builder resolveCanonicalHostname(boolean resolveCanonicalHostname) {
+			this.resolveCanonicalHostname = resolveCanonicalHostname;
+			return this;
+		}
+
+		/**
+		 * Builds the SpnegoAuthProvider instance.
+		 *
+		 * @return a new SpnegoAuthProvider
+		 */
+		public SpnegoAuthProvider build() {
+			return new SpnegoAuthProvider(authenticator, unauthorizedStatusCode, serviceName, resolveCanonicalHostname);
+		}
+	}
```

Comment:
The HTTP client receives the options required for SPNEGO authentication:

1. HTTP status to catch the authentication required error response from the resource server
2. Service principal name
3. Whether to handle canonical hostname